doc/manual/metaprogramming.rst

```diff
@@ -1028,12 +1037,40 @@ copy them*, for the following reasons:
 * the ``foo`` function has side-effects, and it is undefined exactly when,
   how often or how many times these side-effects will occur
 * the ``bar`` function solves a problem that is better solved with multiple
-  dispatch - defining ``bar(x) = x`` and ``bar(x::Integer) = x^2`` will do
+  dispatch - defining ``bar(x) = x`` and ``bar(x::Integer) = x ^ 2`` will do
   the same thing, but it is both simpler and faster.
 * the ``baz`` function is pathologically insane
 
-Instead, now that we have a better understanding for how generated functions
-work, let's use them to build some more advanced functionality...
+Note that the set of operations that should not be attempted in a generated function
+is unbounded, and the runtime system can currently only detect a subset of
+the invalid operations. There are many other operations that will simply
+corrupt the runtime system without notification, usually in subtle
+ways not obviously connected to the bad definition.
+Because the function generator is run during inference,
+it must respect all of the limitations of that code.
+
+Some operations that should not be attempted include:
+
+  1. Caching of native pointers.
+
+  2. Interacting with the contents or methods of Core.Inference in any way.
+
+  3. Observing any mutable state.
+
+     - Inference on the generated function may be run at *any* time,
+       including while your code is attempting to observe or mutate this state.
+
+  4. Taking any locks: C code you call out to may use locks internally,
+     (for example, it is not problematic to call ``malloc``,
+     even though most implementations require locks internally)
+     but don't attempt to hold or acquire any while executing Julia code.
+
+  5. Calling any function that is defined after the body of the generated function.
+     This condition is relaxed for incrementally-loaded precompiled modules to
+     allow calling any function in the module.
+
+Alright, now that we have a better understanding for how generated functions
```

Comment:
for -> of ("now that we have a better understanding of how generated functions work").